db/commitlog/commitlog.cc

```diff
@@ -1566,7 +1563,7 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
 
         if (!cfg.allow_going_over_size_limit && max_disk_size != 0 && totals.total_size_on_disk >= max_disk_size) {
             clogger.debug("Disk usage ({} MB) exceeds maximum ({} MB) - allocation will wait...", totals.total_size_on_disk/(1024*1024), max_disk_size/(1024*1024));
-            auto f = cfg.reuse_segments ? _recycled_segments.not_empty() :  _disk_deletions.get_shared_future();
+            auto f = _recycled_segments.not_empty();
```

Comment:
I remember asking about hints, which don't recycle. /cc @piodul

Reply:
Not actually part of this PR. Just showing because based on #10084. Hints don't recycle, but they don't delete either so the net result is the same - hints will reach shutdown, at which point files are either left for replay, or deleted anyway as we don't (yet?) recycle across CL instances.

Reply:
Calle is right - hints just do not benefit from segment recycling. They do not generate recycled segments and I don't think they would even if they would set cfg.reuse_segments to true (at least not intentionally). They disable the recycling flow just to be explicit that they neither use it nor need it.

If I understand correctly, in order for a segment to become recycled, it must be filled up completely and all references to it (tracked by rp_sets and rp_handles) must be destroyed. Hints manager, after storing a hint and receiving an rp_handle, just calls rp_handle::release which "leaks" the reference, so the reference count never goes to zero and segments cannot be recycled. This is intentional as segments with non-zero reference count are kept on disk when commitlog is destroyed - and hints re-create the commitlog periodically by design.